network/ocsp.go

```diff
@@ -0,0 +1,273 @@
+	if err != nil {
+		return nil, err
+	}
+	ocspURL, err := url.Parse(ocspServerURL)
+	if err != nil {
+		return nil, err
+	}
+	httpRequest.Header.Add("Content-Type", "application/ocsp-request")
+	httpRequest.Header.Add("Accept", "application/ocsp-response")
+	httpRequest.Header.Add("host", ocspURL.Host)
+	httpClient := &http.Client{}
+	httpResponse, err := httpClient.Do(httpRequest)
+	if err != nil {
+		return nil, err
+	}
+	defer httpResponse.Body.Close()
+	output, err := ioutil.ReadAll(httpResponse.Body)
+	if err != nil {
+		return nil, err
+	}
+	ocspResponse, err := ocsp.ParseResponse(output, issuerCert)
+	return ocspResponse, err
+}
+
+// DefaultOCSPVerifier is a default OCSP verifier
+type DefaultOCSPVerifier struct {
+	Config OCSPConfig
+	Client OCSPClient
+}
+
+// ocspServerToCheck is used to plan OCSP requests
+type ocspServerToCheck struct {
+	url      string
+	fromCert bool
+}
+
+// Verify ensures certificate is not revoked by querying configured OCSP servers
+func (v DefaultOCSPVerifier) Verify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	for _, chain := range verifiedChains {
+		log.Debugf("OCSP: Verifying '%s'", chain[0].Subject.String())
+
+		cert := chain[0]
+		issuer := chain[1]
+
+		for _, ocspServer := range cert.OCSPServer {
+			log.Debugf("OCSP: certificate contains OCSP URI: %s", ocspServer)
+		}
+
+		serversToCheck := []ocspServerToCheck{}
+
+		if v.Config.fromCert != ocspFromCertIgnore {
+			for _, ocspServer := range cert.OCSPServer {
+				serverToCheck := ocspServerToCheck{url: ocspServer, fromCert: true}
+				log.Debugf("OCSP: appending server %s, from cert", serverToCheck.url)
+				serversToCheck = append(serversToCheck, serverToCheck)
+			}
+		} else {
+			if len(cert.OCSPServer) > 0 {
+				log.Debugf("OCSP: Ignoring %d OCSP servers from certificate", len(cert.OCSPServer))
+			}
+		}
+
+		if v.Config.url != "" {
+			serverToCheck := ocspServerToCheck{url: v.Config.url, fromCert: false}
+
+			if v.Config.fromCert == ocspFromCertPrefer || v.Config.fromCert == ocspFromCertTrust {
+				log.Debugf("OCSP: appending server %s, from config", serverToCheck.url)
+				serversToCheck = append(serversToCheck, serverToCheck)
+			} else {
+				log.Debugf("OCSP: prepending server %s, from config", serverToCheck.url)
+				serversToCheck = append([]ocspServerToCheck{serverToCheck}, serversToCheck...)
+			}
+		}
+
+		// TODO avoid querying same OCSP more than once
+
+		for _, serverToCheck := range serversToCheck {
+			log.Debugf("OCSP: Trying server %s", serverToCheck.url)
+
+			response, err := v.Client.Query(cert.Issuer.CommonName, cert, issuer, serverToCheck.url)
+			if err != nil {
+				log.WithError(err).Warnf("Cannot query OCSP server at %s", serverToCheck.url)
+
+				if v.Config.required == ocspRequiredAll {
+					return errors.New("Cannot query OCSP server, but --tls_ocsp_required=all was passed")
+				}
+
+				continue
```

Comment:
we should use some counter or flag that at least one check was successful. or we can loop over all values, get error from all of them and not get any valid check.